File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> tmp = new ArrayList<>();
        f(ans,tmp,nums);
        return ans;
    }

    void f(List<List<Integer>> ans,List<Integer> tmp,int[] nums)
    {
        if(tmp.size() == nums.length)
        {
            ans.add(new ArrayList<>(tmp));
            return;
        }
        for(int num: nums)
        {
            if(tmp.contains(num))  continue;
            tmp.add(num);
            f(ans,tmp,nums);
            tmp.remove(tmp.size()-1);
        }
    }
}

File: 0046-permutations/0046-permutations.cpp

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> tmp;
        backtrack(ans, tmp, nums);
        return ans;
    }

private:
    void backtrack(vector<vector<int>>& ans, vector<int>& tmp, vector<int>& nums) {
        if (tmp.size() == nums.size()) {
            ans.push_back(tmp);
            return;
        }
        for (int num : nums) {
            if (find(tmp.begin(), tmp.end(), num) != tmp.end()) continue;
            tmp.push_back(num);
            backtrack(ans, tmp, nums);
            tmp.pop_back();
        }
    }
};
